src/transform/transform.py

```diff
@@ -2,11 +2,135 @@
+	)
+	for addon in addons:
+		for nvdaVersion in NVDAVersions:
+			addonsForVersionChannel = latestAddons[nvdaVersion.apiVer][addon.channel]
+			if (_isAddonCompatible(addon, nvdaVersion) and _isAddonNewer(addonsForVersionChannel, addon)):
+				addonsForVersionChannel[addon.addonId] = addon
+				log.debug(f"added {addon.addonId} {addon.addonVersionNumber.toStr()}")
+			else:
+				log.debug(f"ignoring {addon.addonId} {addon.addonVersionNumber.toStr()}")
+	return latestAddons
+
+
+def writeAddons(addonDir: str, addons: WriteableAddons) -> None:
+	"""
+	Given a unique mapping of (nvdaAPIVersion, channel) -> addon, write the addons to file.
+	Throws a ValidationError and exits if writeable data does not match expected schema.
+	"""
+	for nvdaAPIVersion in addons:
+		for channel in addons[nvdaAPIVersion]:
+			for addonName in addons[nvdaAPIVersion][channel]:
+				addon = addons[nvdaAPIVersion][channel][addonName]
+				addonWritePath = f"{addonDir}/{nvdaAPIVersion.toStr()}/{addonName}"
+				with open(addon.pathToData, "r") as oldAddonFile:
+					addonData = json.load(oldAddonFile)
+				Path(addonWritePath).mkdir(parents=True, exist_ok=True)
+				with open(f"{addonWritePath}/{channel}.json", "w") as newAddonFile:
+					validateJson(addonData, JSONSchemaPaths.ADDON_DATA)
+					json.dump(addonData, newAddonFile)
+
+
+def readAddons(addonDir: str) -> Iterable[Addon]:
+	"""
+	Read addons from a directory and capture required data for processing.
+	Works as a generator to minimize memory usage, as such, each use of iteration should call readAddons.
+	Skips addons and logs errors if the naming schema or json schema do not match what is expected.
+	"""
+	for fileName in glob.glob(f"{addonDir}/**/*.json"):
+		with open(fileName, "r") as addonFile:
+			addonData = json.load(addonFile)
+		try:
+			validateJson(addonData, JSONSchemaPaths.ADDON_DATA)
+		except ValidationError as e:
+			log.error(f"{fileName} doesn't match schema: {e}")
+			continue
+		yield Addon(
+			addonId=addonData["addonId"],
+			addonVersionNumber=AddonVersion.fromDict(addonData["addonVersionNumber"]),
+			pathToData=fileName,
+			channel=addonData["channel"],
+			minNVDAVersion=NVDAVersion.fromDict(addonData["minNVDAVersion"]),
+			lastTestedVersion=NVDAVersion.fromDict(addonData["lastTestedVersion"]),
+		)
+
+
+def readNVDAVersionInfo(pathToFile: str) -> Tuple[VersionCompatibility]:
+	"""
+	Reads and captures NVDA version information from file.
+	"""
+	with open(pathToFile, "r") as NVDAVersionFile:
+		NVDAVersionData = json.load(NVDAVersionFile)
+	validateJson(NVDAVersionData, JSONSchemaPaths.NVDA_VERSIONS)
+	return tuple(
+		VersionCompatibility(
+			nvdaVersion=NVDAVersion.fromStr(version["NVDAVersion"]),
+			apiVer=NVDAVersion.fromStr(version["apiVer"]),
+			backCompatTo=NVDAVersion.fromStr(version["backCompatTo"]),
+		) for version in NVDAVersionData
+	)
+
+
+def emptyDirectory(dir: str) -> None:
+	for filename in glob.glob(dir + "/**/*.json", recursive=True):
```

Comment:
Shouldn't this delete all files, not just json files?
You could consider using `shutil.rmtree(dir)`

Reply:
I'd either expect it to delete all files, or make it the responsibility of the user to specify an empty / non-existent dir. I tend to agree with the safety for devs argument, but in that case check if the folder exists and exit with an error if it does.

Leaving files not generated from the source, could result in non reproducible outcome. The view should be entirely generated from the source. Keeping files means it has "state" of it's own.

Reply:
Done as of 82cef85